Error instead of assertion failure for div by sparse

nairbv opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30260 Error instead of assertion failure for div by sparse**

fixes: https://github.com/pytorch/pytorch/issues/30044

Summary:

Without this PR,

```
>>> torch.tensor(1.) / torch.tensor(1.).to_sparse()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: r.is_sparse() INTERNAL ASSERT FAILED at /Users/distiller/project/conda/conda-bld/pytorch_1570710797334/work/aten/src/ATen/native/sparse/SparseTensorMath.cpp:168, please report a bug to PyTorch.
```

Test Plan:

Ran the same code with this change:

```
In [1]: import torch
In [2]: torch.tensor(1).to_sparse() / torch.tensor(1).to_sparse()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-7177f54f30bb> in <module>
----> 1 torch.tensor(1).to_sparse() / torch.tensor(1).to_sparse()

RuntimeError: Unsupported tensor layout
```

Differential Revision: [D18657387](https://our.internmc.facebook.com/intern/diff/D18657387)

nairbv committed:
Error instead of assertion failure for div by sparse

fixes: https://github.com/pytorch/pytorch/issues/30044

Summary:

Without this PR,

```
>>> torch.tensor(1.) / torch.tensor(1.).to_sparse()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: r.is_sparse() INTERNAL ASSERT FAILED at /Users/distiller/project/conda/conda-bld/pytorch_1570710797334/work/aten/src/ATen/native/sparse/SparseTensorMath.cpp:168, please report a bug to PyTorch.
```

Test Plan:

Ran the same code with this change:

```
In [1]: import torch
In [2]: torch.tensor(1).to_sparse() / torch.tensor(1).to_sparse()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-7177f54f30bb> in <module>
----> 1 torch.tensor(1).to_sparse() / torch.tensor(1).to_sparse()

RuntimeError: Unsupported tensor layout
```

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -164,6 +164,7 @@ Tensor& div_sparse_(Tensor& self, const Tensor& value) {
 SparseTensor& div_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, const Tensor& value) {
   TORCH_CHECK(value.dim() == 0, "sparse division only supports division by a scalar (got shape ",
       value.sizes(), " for argument 'other')");
+  TORCH_CHECK(!value.is_sparse(), "Unsupported tensor layout");
```
It's good to tell the user what exactly is unsupported and what is unsupported. Judging by this, it seems like "sparse / sparse" is unsupported, but "sparse / dense" is supported. Is that right?

Also, this needs a test

nairbv replied:
added, though I feel a bit silly adding a test just to validate the particular text of an error.

zou3519 replied:
Well, someone could come along and delete the error message and we'd never know. I think that having tests to validate that error messages exist and are what we expect is valuable to have.

zou3519 reviewed (changes requested):
Needs test. Also it would be good to improve the error message

nairbv committed:
Update on "Error instead of assertion failure for div by sparse"


fixes: https://github.com/pytorch/pytorch/issues/30044

Summary:

Without this PR,

```
>>> torch.tensor(1.) / torch.tensor(1.).to_sparse()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: r.is_sparse() INTERNAL ASSERT FAILED at /Users/distiller/project/conda/conda-bld/pytorch_1570710797334/work/aten/src/ATen/native/sparse/SparseTensorMath.cpp:168, please report a bug to PyTorch.
```

Test Plan:

Ran the same code with this change:

```
In [1]: import torch
In [2]: torch.tensor(1).to_sparse() / torch.tensor(1).to_sparse()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-7177f54f30bb> in <module>
----> 1 torch.tensor(1).to_sparse() / torch.tensor(1).to_sparse()

RuntimeError: Unsupported tensor layout
```

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/sparse/SparseTensorMath.cpp:
```diff
@@ -164,6 +164,7 @@ Tensor& div_sparse_(Tensor& self, const Tensor& value) {
 SparseTensor& div_out_sparse_zerodim(SparseTensor& r, const SparseTensor& t, const Tensor& value) {
   TORCH_CHECK(value.dim() == 0, "sparse division only supports division by a scalar (got shape ",
       value.sizes(), " for argument 'other')");
+  TORCH_CHECK(!value.is_sparse(), "Unsupported tensor layout: Can't divide a sparse tensor by a sparse value");
```
nit: "sparse value" sounds like it is a scalar, not a tensor. Furthermore, users won't really know that the second argument to `div` is called `value`. Could we just say "can't divide a sparse tensor by a sparse tensor"?

nairbv replied:
it is sort of supposed to be a "value" though ... you can't pass a dimensioned tensor of any layout (as asserted above this line)

zou3519 replied:
Ah, my bad, I missed that.

nairbv replied:
updated again to be clearer anyway

zou3519 commented on test/test_sparse.py:
```diff
@@ -2093,6 +2093,10 @@ def test_pickle(self):
+    def test_div_by_sparse_error(self):
+
+        self.assertRaisesRegex(RuntimeError, 'Unsupported tensor layout', 
+                               lambda: torch.tensor(1., device=self.device) / torch.tensor(1., device=self.device).to_sparse())
```
I don't think this is right. Shouldn't we be testing sparse / sparse division? (this is dense / sparse division right now).

Also, `torch.zeros(1, layout = torch.sparse, device=torch.device)` is a faster way to create sparse tensors.

nairbv replied:
ah I must have pasted wrong. good to know about the `layout=` option

nairbv replied:
Any idea what I'm doing wrong here?

```
In [22]: torch.ones(())
Out[22]: tensor(1.)

In [23]: torch.ones((), layout=torch.sparse)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-ed8e3fb1ad17> in <module>
----> 1 torch.ones((), layout=torch.sparse)

TypeError: ones() received an invalid combination of arguments - got (tuple, layout=module), but expected one of:
 * (tuple of ints size, tuple of names names, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
 * (tuple of ints size, Tensor out, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
```

zou3519 replied:
The name of the layout is actually `torch.sparse_coo` ...

nairbv replied:
`full(...) is not implemented for sparse layout` ... `zeros` works but it seems a bit awkward where the test is division.

nairbv committed:
Update on "Error instead of assertion failure for div by sparse"


fixes: https://github.com/pytorch/pytorch/issues/30044

Summary:

Without this PR,

```
>>> torch.tensor(1.) / torch.tensor(1.).to_sparse()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: r.is_sparse() INTERNAL ASSERT FAILED at /Users/distiller/project/conda/conda-bld/pytorch_1570710797334/work/aten/src/ATen/native/sparse/SparseTensorMath.cpp:168, please report a bug to PyTorch.
```

Test Plan:

Ran the same code with this change:

```
In [1]: import torch
In [2]: torch.tensor(1).to_sparse() / torch.tensor(1).to_sparse()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-2-7177f54f30bb> in <module>
----> 1 torch.tensor(1).to_sparse() / torch.tensor(1).to_sparse()

RuntimeError: Unsupported tensor layout
```

[ghstack-poisoned]
